### PR TITLE
Gun attachments

### PR DESCRIPTION
**Pull request by FalloutFalcon**
## About The Pull Request
Finishing #145
Requires #2877

## Changelog

:cl:
refactor: refactored attachments to be modular
/:cl:

### CODE REVIEW

File: code/datums/components/attachment.dm [hunk 1]
```diff
@@ -0,0 +1,175 @@
+/datum/component/attachment
```
Comment (thgvr) [inline code/datums/components/attachment.dm:1]: Autodoc all of this please

File: code/datums/components/attachment.dm [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	return ..()
+
+/datum/component/attachment/proc/try_toggle(obj/item/parent, obj/item/holder, mob/user)
+	SIGNAL_HANDLER
+	if(attach_features_flags & ATTACH_TOGGLE)
+		INVOKE_ASYNC(src, PROC_REF(do_toggle), parent, holder, user)
+		holder.update_icon()
+		attachment_toggle_action.UpdateButtonIcon()
+
+
```
Comment (thgvr) [inline code/datums/components/attachment.dm]: extra newline here

File: code/datums/components/attachment.dm [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+/datum/component/attachment/proc/send_slot(obj/item/parent)
+	SIGNAL_HANDLER
+	return attachment_slot_to_bflag(slot)
+
+/datum/action/attachment
+	name = "Toggle Attachment"
+	check_flags = AB_CHECK_HANDS_BLOCKED|AB_CHECK_CONSCIOUS
+	button_icon_state = null
+	var/obj/item/gun/gun = null
```
Comment (thgvr) [inline code/datums/components/attachment.dm:144]: autodoc

File: code/datums/components/attachment_holder.dm [hunk 4]
```diff
@@ -0,0 +1,185 @@
+/datum/component/attachment_holder
```
Comment (thgvr) [inline code/datums/components/attachment_holder.dm:1]: autodoc, yet again

File: code/game/objects/items/attachments/_attachment.dm [hunk 5]
```diff
@@ -0,0 +1,88 @@
+/obj/item/attachment
```
Comment (thgvr) [inline code/game/objects/items/attachments/_attachment.dm:2]: Some of these vars are autodoc and the rest aren't, ditto

File: code/game/objects/items/attachments/_attachment.dm [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	playsound(user, toggled ? toggle_on_sound : toggle_off_sound, 40, TRUE)
+	toggled = !toggled
+	icon_state = "[initial(icon_state)][toggled ? "-on" : ""]"
+
+/// Checks if a user should be allowed to attach this attachment to the given parent
+/obj/item/attachment/proc/Attach(obj/item/gun/gun, mob/user)
+	SHOULD_CALL_PARENT(TRUE)
+
+	if(toggled)
+		to_chat(user, "<span class='warning'>You cannot attach [src] while it is active!</span>")
```
Comment (thgvr) [inline code/game/objects/items/attachments/_attachment.dm]: Span macro

File: code/game/objects/items/attachments/_attachment.dm [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+/obj/item/attachment/proc/Toggle(obj/item/gun/gun, mob/user)
+	SHOULD_CALL_PARENT(TRUE)
+
+	playsound(user, toggled ? toggle_on_sound : toggle_off_sound, 40, TRUE)
+	toggled = !toggled
+	icon_state = "[initial(icon_state)][toggled ? "-on" : ""]"
+
+/// Checks if a user should be allowed to attach this attachment to the given parent
+/obj/item/attachment/proc/Attach(obj/item/gun/gun, mob/user)
```
Comment (thgvr) [inline code/game/objects/items/attachments/_attachment.dm]: Proc name needs to be better and in the right case - 

`add_attachment`?
`apply_attachment`

File: code/game/objects/items/attachments/_attachment.dm [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+		CALLBACK(src, PROC_REF(Toggle)), \
+		CALLBACK(src, PROC_REF(PreAttack)), \
+		signals)
+
+/obj/item/attachment/Destroy()
+	qdel(attachment_comp)
+	attachment_comp = null
+	. = ..()
+
+/obj/item/attachment/proc/Toggle(obj/item/gun/gun, mob/user)
```
Comment (thgvr) [inline code/game/objects/items/attachments/_attachment.dm]: Proc name needs to be better and in the right case -

`toggle_attachment`

File: code/game/objects/items/attachments/_attachment.dm [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	SHOULD_CALL_PARENT(TRUE)
+
+	if(toggled)
+		Toggle(gun, user)
+
+	apply_modifiers(gun, user, FALSE)
+	playsound(src.loc, 'sound/weapons/gun/pistol/mag_release_alt.ogg', 75, 1)
+	return TRUE
+
+/obj/item/attachment/proc/PreAttack(obj/item/gun/gun, atom/target, mob/user, list/params)
```
Comment (thgvr) [inline code/game/objects/items/attachments/_attachment.dm]: same, case and better name.

File: code/game/objects/items/attachments/_attachment.dm [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+	if(toggled)
+		to_chat(user, "<span class='warning'>You cannot attach [src] while it is active!</span>")
+		return FALSE
+
+	apply_modifiers(gun, user, TRUE)
+	playsound(src.loc, 'sound/weapons/gun/pistol/mag_insert_alt.ogg', 75, 1)
+	return TRUE
+
+/obj/item/attachment/proc/Detach(obj/item/gun/gun, mob/user)
```
Comment (thgvr) [inline code/game/objects/items/attachments/_attachment.dm]: `remove_attachment` maybe

File: code/game/objects/items/attachments/bayonet.dm [hunk 11]
```diff
@@ -0,0 +1,22 @@
+/obj/item/attachment/bayonet
+	name = "bayonet"
+	desc = "Stabby-Stabby"
```
Comment (thgvr) [inline code/game/objects/items/attachments/bayonet.dm:3]: Is this ever even seeable? Just have no description if it is.

File: code/game/objects/items/attachments/laser_sight.dm [hunk 12]
```diff
@@ -0,0 +1,23 @@
+/obj/item/attachment/laser_sight
+	name = "laser sight"
+	desc = "This laser sight is designed to be rail-mounted on a compatible firearm to provide increased accuracy."
```
Comment (thgvr) [inline code/game/objects/items/attachments/laser_sight.dm]:
```suggestion
	desc = "Designed to be rail-mounted on a compatible firearm to provide increased accuracy and decreased spread."
```

File: code/game/objects/items/attachments/laser_sight.dm [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+	if(toggled)
+		gun.spread -= 3
+		gun.spread_unwielded -= 3
+		gun.wield_delay -= 0.3 SECONDS
+	else
+		gun.spread += 3
+		gun.spread_unwielded += 3
+		gun.wield_delay += 0.3 SECONDS
+
+	playsound(user, toggled ? 'sound/weapons/magin.ogg' : 'sound/weapons/magout.ogg', 40, TRUE)
```
Comment (thgvr) [inline code/game/objects/items/attachments/laser_sight.dm:23]: If these are used in more than one spot it needs to be a define.

File: code/game/objects/items/attachments/rail_light.dm [hunk 14]
```diff
@@ -0,0 +1,33 @@
+/obj/item/attachment/rail_light
+	name = "rail light"
+	desc = "Rail mounted gun light for better visibility down range."
```
Comment (thgvr) [inline code/game/objects/items/attachments/rail_light.dm]:
```suggestion
	desc = "A flashlight made to be mounted on a firearm."
```

File: code/game/objects/items/attachments/rail_light.dm [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+	icon_state = "raillight"
+	light_color = "#FFCC66"
```
Comment (thgvr) [inline code/game/objects/items/attachments/rail_light.dm]: color define?

File: code/game/objects/items/attachments/silencer.dm [hunk 16]
```diff
@@ -0,0 +1,19 @@
+/obj/item/attachment/silencer
+	name = "silencer"
+	desc = "For when you need to kill someone but not be seen killing someone!"
```
Comment (thgvr) [inline code/game/objects/items/attachments/silencer.dm]:
```suggestion
	desc = "An attachment for the barrel of a firearm. Muffles the gunshot and muzzle flash."
```

File: code/game/objects/items/attachments/silencer.dm [hunk 17]
```diff
@@ -0,0 +1,19 @@
+/obj/item/attachment/silencer
+	name = "silencer"
```
Comment (thgvr) [inline code/game/objects/items/attachments/silencer.dm]:
```suggestion
	name = "suppressor"
```

File: code/game/objects/items/attachments/stock.dm [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+		gun.w_class -= 1
+		gun.wield_delay -= toggled_wield_delay
+		gun.wield_slowdown -= toggled_slowdown
+		gun.recoil -= toggled_recoil_bonus
+		gun.spread -= toggled_spread_bonus
+
+	if(gun.wielded)
+		user.add_or_update_variable_movespeed_modifier(/datum/movespeed_modifier/gun, multiplicative_slowdown = gun.wield_slowdown)
+
+	playsound(src, 'sound/weapons/empty.ogg', 100, 1)
```
Comment (thgvr) [inline code/game/objects/items/attachments/stock.dm]: Define when used in multiple places

File: code/game/objects/items/attachments/stock.dm [hunk 19]
```diff
@@ -0,0 +1,40 @@
+/obj/item/attachment/foldable_stock
+	name = "foldable stock"
```
Comment (thgvr) [inline code/game/objects/items/attachments/stock.dm]:
```suggestion
	name = "folding stock"
```

File: code/game/objects/items/attachments/stock.dm [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+	desc = "A foldable stock that can be attached to a weapon to improve stability."
```
Comment (thgvr) [inline code/game/objects/items/attachments/stock.dm]:
```suggestion
	desc = "A folding stock that can be attached to certain weapons to improve stability and decreases recoil."
```

File: code/game/objects/items/attachments/wip.dm [hunk 21]
```diff
@@ -0,0 +1,113 @@
+/*
+/obj/item/attachment/bipod
```
Comment (thgvr) [inline code/game/objects/items/attachments/wip.dm]: don't leave commented files in

File: code/modules/cargo/packs/gun.dm [hunk 22]
```diff
@@ -189,3 +189,31 @@
 	cost = 5000
 	contains = list(/obj/item/storage/guncase/skm)
 	crate_name = "auto rifle crate"
+
+/datum/supply_pack/gun/attachment/rail_light
+	name = "Tactical Rail Light Crate"
+	desc = "Rail mounted gun light for better visibility down range."
```
Comment (thgvr) [inline code/modules/cargo/packs/gun.dm]:
```suggestion
	desc = "Contains a single rail light to be mounted on a firearm."
```

File: code/modules/cargo/packs/gun.dm [hunk 23]
```diff
@@ -189,3 +189,31 @@
...
+	cost = 250
+	contains = list(/obj/item/attachment/rail_light)
+	crate_name = "rail light crate"
+
+/datum/supply_pack/gun/attachment/laser_sight
+	name = "Laser Sight Crate"
+	desc = "Laser sight for better accuracy down range."
```
Comment (thgvr) [inline code/modules/cargo/packs/gun.dm]:
```suggestion
	desc = "Contains a single rail light to be mounted on a firearm."
```

File: code/modules/cargo/packs/gun.dm [hunk 24]
```diff
@@ -189,3 +189,31 @@
...
+	cost = 250
+	contains = list(/obj/item/attachment/laser_sight)
+	crate_name = "laser sight crate"
+
+/datum/supply_pack/gun/attachment/bayonet
+	name = "Bayonet Crate"
+	desc = "Bayonet for close quarters combat."
```
Comment (thgvr) [inline code/modules/cargo/packs/gun.dm]:
```suggestion
	desc = "Contains a single bayonet to be mounted on a firearm."
```

File: code/modules/cargo/packs/gun.dm [hunk 25]
```diff
@@ -189,3 +189,31 @@
...
+	cost = 250
+	contains = list(/obj/item/attachment/bayonet)
+	crate_name = "bayonet crate"
+
+/datum/supply_pack/gun/attachment/silencer
+	name = "Silencer Crate"
```
Comment (thgvr) [inline code/modules/cargo/packs/gun.dm]:
```suggestion
	name = "Suppressor Crate"
```

File: code/modules/cargo/packs/gun.dm [hunk 26]
```diff
@@ -189,3 +189,31 @@
...
+	desc = "Silencer for stealthy operations."
```
Comment (thgvr) [inline code/modules/cargo/packs/gun.dm]:
```suggestion
	desc = "Contains a single suppressor to be mounted on a firearm."
```

File: code/modules/mob/living/simple_animal/hostile/megafauna/codename_claw.dm [hunk 27]
```diff
@@ -226,15 +223,6 @@
 	empulse(src, 2, 4)
 	shouldnt_move = FALSE
 
-/////TENTACLE
-/mob/living/simple_animal/hostile/megafauna/claw/proc/tentacle(target)
```
Comment (thgvr) [inline code/modules/mob/living/simple_animal/hostile/megafauna/codename_claw.dm]: Why changes in this file?

File: code/modules/projectiles/gun.dm [hunk 28]
```diff
@@ -4,6 +4,8 @@
 	icon = 'icons/obj/guns/projectile.dmi'
 	icon_state = "flatgun"
 	item_state = "gun"
+	lefthand_file = 'icons/mob/inhands/weapons/guns_lefthand.dmi'
+	righthand_file = 'icons/mob/inhands/weapons/guns_righthand.dmi'
```
Comment (thgvr) [inline code/modules/projectiles/gun.dm]: Might be worth making a define here if the same dmi is used a lot.

File: code/modules/projectiles/gun.dm [hunk 29]
```diff
@@ -121,43 +207,126 @@
...
+	//Zoom toggle
+	var/zoomed = FALSE
+	///Distance in TURFs to move the user's screen forward (the "zoom" effect)
+	var/zoom_amt = 3
+	var/zoom_out_amt = 0
+	var/datum/action/toggle_scope_zoom/azoom
 
-	//gun saftey
+/*
+ * Saftey
```
Comment (thgvr) [inline code/modules/projectiles/gun.dm]:
```suggestion
 * Safety
```

**Review by thgvr — DISMISSED** [review]
Some things, nitpicks, code styling, weird bits

Please remember newlines at the end of your files - it's missing a lot and the linter really doesn't like it

Autodoc all of the attachment stuff as well, left comments.

Could be worth making an explanation/guide at the top of the attachments on what everything does and how to make a new one correctly.

File: code/modules/mob/living/simple_animal/hostile/megafauna/codename_claw.dm (hunk 27 above)
Comment (FalloutFalcon) [reply code/modules/mob/living/simple_animal/hostile/megafauna/codename_claw.dm]: oops, merge accident i think

**Pull request merged**